File: packages/observing-api/src/routes/identifications.ts
/**
 * Identification routes - read and write endpoints
 */

import { Router } from "express";
import { Database, CommunityIdCalculator, TaxonomyClient } from "observing-shared";
import { enrichIdentifications } from "../enrichment.js";
import { logger } from "../middleware/logging.js";
import { requireAuth } from "../middleware/auth.js";
import { InternalClient } from "../internal-client.js";

export function createIdentificationRoutes(
  db: Database,
  communityId: CommunityIdCalculator,
  taxonomy: TaxonomyClient,
  internalClient: InternalClient
): Router {
  const router = Router();

  // Create a new identification
  router.post("/", requireAuth, async (req, res) => {
    try {
      const sessionDid = req.user!.did;
      const {
        occurrenceUri,
        occurrenceCid,
        subjectIndex = 0,
        taxonName,
        taxonRank = "species",
        comment,
        isAgreement = false,
        confidence = "medium",
      } = req.body;

      // Validate required fields
      if (!occurrenceUri || !occurrenceCid) {
        res.status(400).json({ error: "occurrenceUri and occurrenceCid are required" });
        return;
      }

      if (!taxonName || taxonName.trim().length === 0) {
        res.status(400).json({ error: "taxonName is required" });
        return;
      }

      if (taxonName.length > 256) {
        res.status(400).json({ error: "taxonName too long (max 256 characters)" });
        return;
      }

      if (comment && comment.length > 3000) {
        res.status(400).json({ error: "comment too long (max 3000 characters)" });
        return;
      }

      // Fetch taxonomy hierarchy from GBIF
      const validationResult = await taxonomy.validate(taxonName.trim());
      const taxon = validationResult.taxon;

      // Build the identification record
      const record = {
        $type: "org.rwell.test.identification",
        subject: {
          uri: occurrenceUri,
          cid: occurrenceCid,
        },
        subjectIndex,
        taxonName: taxonName.trim(),
        taxonRank,
        comment: comment?.trim() || undefined,
        isAgreement,
        confidence,
        createdAt: new Date().toISOString(),
        taxonId: taxon?.id,
        vernacularName: taxon?.commonName,
        kingdom: taxon?.kingdom,
        phylum: taxon?.phylum,
        class: taxon?.class,
        order: taxon?.order,
        family: taxon?.family,
        genus: taxon?.genus,
      };

      // Create the record via internal RPC
      const result = await internalClient.createRecord(
        sessionDid,
        "org.rwell.test.identification",
        record
      );

      if (!result.success || !result.uri) {
        res.status(500).json({ error: result.error || "Failed to create identification" });
        return;
      }

      logger.info({ uri: result.uri, occurrenceUri, isAgreement }, "Created identification via internal RPC");

      res.status(201).json({
        success: true,
        uri: result.uri,
        cid: result.cid,
      });
    } catch (error) {
      logger.error({ err: error }, "Error creating identification");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Delete an identification
  router.delete("/:uri(*)", requireAuth, async (req, res) => {
    try {
      const uri = req.params["uri"];
      if (!uri) {
        res.status(400).json({ error: "uri is required" });
        return;
      }

      // Parse AT URI: at://did:plc:xxx/org.rwell.test.identification/rkey
      const uriMatch = uri.match(/^at:\/\/([^/]+)\/([^/]+)\/([^/]+)$/);
      if (!uriMatch) {
        res.status(400).json({ error: "Invalid AT URI format" });
        return;
      }

      const recordDid = uriMatch[1]!;
      const collection = uriMatch[2]!;
      const rkey = uriMatch[3]!;

      const sessionDid = req.user!.did;

      // Verify user owns this identification
      if (sessionDid !== recordDid) {
        res.status(403).json({ error: "You can only delete your own identifications" });
        return;
      }

      // Delete from AT Protocol via internal RPC
      const result = await internalClient.deleteRecord(sessionDid, collection, rkey);
      if (!result.success) {
        res.status(500).json({ error: result.error || "Failed to delete from AT Protocol" });
        return;
      }

      // Delete from database (also refreshes community IDs)
      await db.deleteIdentification(uri);

      logger.info({ uri, sessionDid }, "Deleted identification");

      res.json({ success: true });
    } catch (error) {
      logger.error({ err: error }, "Error deleting identification");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get identifications for an occurrence
  router.get("/:occurrenceUri(*)", async (req, res) => {
    try {
      const occurrenceUri = req.params["occurrenceUri"];
      if (!occurrenceUri) {
        res.status(400).json({ error: "occurrenceUri is required" });
        return;
      }
      const rows = await db.getIdentificationsForOccurrence(occurrenceUri);
      const identifications = await enrichIdentifications(rows);

      // Calculate community ID
      const communityTaxon = await communityId.calculate(occurrenceUri);

      res.json({
        identifications,
        communityId: communityTaxon,
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching identifications");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  return router;
}

File: crates/observing-appview/src/routes/identifications.rs
use axum::extract::{Path, State};
use axum::Json;
use jacquard_common::types::collection::Collection;
use observing_lexicons::bio_lexicons::temp::identification::{
    Identification, IdentificationRecord, IdentificationTaxonRank,
};
use serde::Deserialize;
use tracing::{info, warn};
use ts_rs::TS;

use crate::auth::{self, AuthUser};
use crate::constants;
use crate::enrichment;
use crate::error::AppError;
use crate::responses::{IdentificationListResponse, RecordCreatedResponse, SuccessResponse};
use crate::state::AppState;
use crate::taxonomy_client::TaxonFields;
use crate::validation::validate_string_length;
use at_uri_parser::AtUri;

pub async fn get_for_occurrence(
    State(state): State<AppState>,
    Path(occurrence_uri): Path<String>,
) -> Result<Json<IdentificationListResponse>, AppError> {
    let rows =
        observing_db::identifications::get_for_occurrence(&state.pool, &occurrence_uri).await?;

    let identifications = enrichment::enrich_identifications(&state.resolver, &rows).await;

    let community_id =
        observing_db::identifications::get_community_id(&state.pool, &occurrence_uri).await?;

    Ok(Json(IdentificationListResponse {
        identifications,
        community_id,
    }))
}

// --- Write handlers ---

#[derive(Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "bindings/")]
pub struct CreateIdentificationRequest {
    occurrence_uri: String,
    occurrence_cid: String,
    scientific_name: String,
    #[ts(optional)]
    taxon_rank: Option<String>,
    #[ts(optional)]
    is_agreement: Option<bool>,
}

pub async fn create_identification(
    State(state): State<AppState>,
    user: AuthUser,
    Json(body): Json<CreateIdentificationRequest>,
) -> Result<Json<RecordCreatedResponse>, AppError> {
    validate_string_length(
        &body.scientific_name,
        1,
        constants::MAX_SCIENTIFIC_NAME_LENGTH,
        "Scientific name",
    )?;

    // Validate taxonomy via GBIF
    let fields = TaxonFields::from_validation(
        &state.taxonomy,
        &body.scientific_name,
        body.taxon_rank.clone(),
    )
    .await;

    let occurrence = auth::build_strong_ref(&body.occurrence_uri, &body.occurrence_cid)?;

    let record = Identification::new()
        .occurrence(occurrence)
        .scientific_name(&*body.scientific_name)
        .maybe_taxon_rank(
            fields
                .taxon_rank
                .as_deref()
                .map(|s| IdentificationTaxonRank::from_value(s.into())),
        )
        .maybe_kingdom(fields.kingdom.as_deref().map(Into::into))
        .build();

    let mut record_value = auth::serialize_at_record(&record)?;

    // App-specific fields (not in upstream lexicon, stored as extra data in the AT Protocol record)
    if let Some(obj) = record_value.as_object_mut() {
        obj.insert(
            "createdAt".to_string(),
            serde_json::json!(chrono::Utc::now().to_rfc3339()),
        );
        obj.insert(
            "isAgreement".to_string(),
            serde_json::json!(body.is_agreement.unwrap_or(false)),
        );
    }

    let (agent, did_parsed) = auth::require_agent(&state.oauth_client, &user.did).await?;
    let resp = auth::create_at_record(&agent, did_parsed, IdentificationRecord::NSID, record_value)
        .await?;

    info!(uri = %resp.uri, "Created identification");

    Ok(Json(RecordCreatedResponse {
        success: true,
        uri: resp.uri.to_string(),
        cid: resp.cid.as_ref().to_string(),
    }))
}

pub async fn delete_identification(
    State(state): State<AppState>,
    user: AuthUser,
    Path(uri): Path<String>,
) -> Result<Json<SuccessResponse>, AppError> {
    let at_uri = AtUri::parse(&uri).ok_or_else(|| AppError::BadRequest("Invalid AT URI".into()))?;

    if at_uri.did != user.did {
        return Err(AppError::Forbidden(
            "You can only delete your own records".into(),
        ));
    }

    let (agent, did_parsed) = auth::require_agent(&state.oauth_client, &user.did).await?;
    agent
        .api
        .com
        .atproto
        .repo
        .delete_record(
            atrium_api::com::atproto::repo::delete_record::InputData {
                collection: at_uri
                    .collection
                    .parse()
                    .map_err(|e| AppError::Internal(format!("Invalid collection: {e}")))?,
                repo: atrium_api::types::string::AtIdentifier::Did(did_parsed),
                rkey: at_uri
                    .rkey
                    .parse()
                    .map_err(|e| AppError::Internal(format!("Invalid rkey: {e}")))?,
                swap_commit: None,
                swap_record: None,
            }
            .into(),
        )
        .await
        .map_err(|e| {
            if matches!(e, atrium_api::xrpc::Error::Authentication(_)) {
                tracing::warn!(error = %e, "AT Protocol authentication failed (session expired)");
                AppError::Unauthorized
            } else {
                AppError::Internal(format!("Failed to delete record: {e}"))
            }
        })?;

    // Delete from local DB (refreshes community IDs)
    if let Err(e) = observing_db::identifications::delete(&state.pool, &uri).await {
        warn!(error = %e, "Failed to delete identification from local DB");
    }

    Ok(Json(SuccessResponse { success: true }))
}
